### ISSUE TITLE
Readmodel DomainEvent with list of entities

### ISSUE THREAD

**Issue by wgtmpeters** [eventflow/EventFlow#502]
In all examples so far I only see readmodel with simple properties.
How would you normalize a domainevent that contains for example a list of entities that are part of the Aggregate?
I would expect those to end up like separate readmodels, one for each entity, however we only have one DomainEvent.
The queryhandlers would return only one by entity id or all entities for a specific aggregate Identity.

**Comment by frankebersoll** [comment]
Read Models and Aggregates are different things. You can use them 1:1, but there might be cases where you have many read models for a single aggregate instance or one read model instance that receives events of different aggregates. 

Read model registration methods have an overload with an additional generic parameter, where you can specify an `IReadModelLocator` implementation. This defines only one method: `IEnumerable<string>  GetReadModelIds(IDomainEvent domainEvent)` where you project a single event to any number of read model IDs (which are just strings). The framework will create or load an instance of the read model for each ID and apply the event.

**Comment by wgtmpeters** [comment]
That is what I understand from the example but the example in the documentation is actually simply about an event with UserNicknameAdded which will only yield 1 ReadmodelId back.

But what about the example where a domain events has a list of Entities and I yield return each of those.
I can understand that for each a new readmodel is created and applied with the event but how do I then know which Entity from the DomainEvent to use to update the Readmodel?

Would a ReadModel not always need some identity property ? After reading in the code a bit that seems to be the case when looking at specific implementations of the ReadModelStores.

So a readmodel should always have a property which will be set by the readmodel store before applying the DomainEvent.
This is not really explained that clearly in the documentation by my opinion. Maybe an example that is about the use case where 1 domain event results in multiple readmodels would be a nice addition :).

@frankebersoll  thank you for the explanation and for your work with the EF Readstore, hope I have some time soon to experiment with it.

**Comment by frankebersoll** [comment]
@wgtmpeters Can you give an example? I think it's easier to reason about when we have a concrete example.

@rasmus Is there a reason why we set the read model ID only after applying events? I know of at least one application of ours that actually makes assumptions that ID == default in the Apply method means "I am a new read model" while ID != default means "I am being updated".

**Comment by wgtmpeters** [comment]
In the existing code this is used for a event ThingyMessageAdded Event and we now assume we have a event hat adds multiple messages at once.
Ok lets assume we have a new DomainEvent for the ThingyAggregate and a locator.
And a SQLiteThingyMessageReadModel.

```javascript
public class ThingyMessagesAddedEvent : AggregateEvent<ThingyAggregate, ThingyId>
    {
        public ThingyMessagesAddedEvent(
            List<ThingyMessage> thingyMessages)
        {
            ThingyMessages = thingyMessages;
        }

        public List<ThingyMessage> ThingyMessages { get; }
    }
public class ThingyMessagesLocator : IReadModelLocator
    {
        public IEnumerable<string> GetReadModelIds(IDomainEvent domainEvent)
        {
            var messagesAddedEvent = domainEvent as IDomainEvent<ThingyAggregate, ThingyId, ThingyMessagesAddedEvent>;
            if (messagesAddedEvent == null)
            {
                yield break;
            }
           foreach(var thingyMessage in messagesAddedEvent.AggregateEvent.ThingyMessages) 
           {
               yield return thingyMessage.Id.Value;
           }
        }
    }

[Table("ReadModel-ThingyMessage")]
    public class SQLiteThingyMessageReadModel : IReadModel,
        IAmReadModelFor<ThingyAggregate, ThingyId, ThingyMessagesAddedEvent>
    {
        public string ThingyId { get; set; }

        [SqlReadModelIdentityColumn]
        public string MessageId { get; set; }

        public string Message { get; set; }

        public void Apply(IReadModelContext context, IDomainEvent<ThingyAggregate, ThingyId, ThingyMessagesAddedEvent> domainEvent)
        {
            ThingyId = domainEvent.AggregateIdentity.Value;
            // this can only be done when the created readmodel is updated with its identity before applying the event.
            Message = domainEvent.AggregateEvent.ThingyMessages.Single(m=>m.Id.Value == MessageId).Message;
        }

        public ThingyMessage ToThingyMessage()
        {
            return new ThingyMessage(
                ThingyMessageId.With(MessageId),
                Message);
        }
    }
```
I would then have to assume that the MessageId is already set when the event is applied.
So far it looks like this depends on the actual implementation of the Readstore is this works.
Some do seem to update the newly created readmodel with the readModelId.

For example from the MssqlReadmodelstore
```javascript
  if (readModel == null)
            {
                readModel = await _readModelFactory.CreateAsync(readModelUpdate.ReadModelId, cancellationToken)
                    .ConfigureAwait(false);
                mssqlReadModel = readModel as IMssqlReadModel;
                if (mssqlReadModel != null)
                {
                    mssqlReadModel.AggregateId = readModelUpdate.ReadModelId;
                    mssqlReadModel.CreateTime = readModelUpdate.DomainEvents.First().Timestamp;
                }
```

have not tested this though but I think it should always be the case that the key column of the Readmodel is always set before applying the event.
Maybe the Factory.Create method should somehow make sure that is the case, it now simply creates and only logs the ID that is given as a parameter.
If we could extend that factory to always look for a constructor with 1 string parameter for the readmodelid then that can be set within the constructor of the readmodel itself.
The other way would be that all implementations of the readmodelstores should make sure the readmodelId is set the Identity property of the Readmodel.

**Comment by frankebersoll** [comment]
I have created PR #506 which adds new properties to the `ReadModelContext`, this should help in your scenario. Setting an ID property on the read model itself (regardless of before or after applying events) seems to be an implementation detail of the respective store - the simplest store (InMemory) doesn't even do it, as it uses external keys in a dictionary.

**Comment by wgtmpeters** [comment]
Makes sense that it is part of the injected context. It should be the same for all, not depending on the readstore that is used.

This will help, maybe we should use this example in the documentation as well.

**Comment by rasmus** [comment]
@wgtmpeters and @frankebersoll thanks for the analysis. Any differences in the implementations is simply because I forgot or didn't think of. The read model part of EventFlow could do with a thorough inspection as we at eBay only use the functionality very limited and thus isn't something I use that often.

**Comment by wgtmpeters** [comment]
was fixed by the new commits in #506
